File: orb/apps/android/kagami-android/app/src/main/java/com/kagami/android/hub/HubManager.kt
package com.kagami.android.hub

import android.content.Context
import android.net.nsd.NsdManager
import android.net.nsd.NsdServiceInfo
import android.util.Log
import kotlinx.coroutines.*
import kotlinx.coroutines.flow.*
import kotlinx.serialization.Serializable
import kotlinx.serialization.json.Json
import okhttp3.*
import okhttp3.MediaType.Companion.toMediaType
import okhttp3.RequestBody.Companion.toRequestBody
import java.io.IOException
import java.util.concurrent.TimeUnit
import javax.inject.Inject
import javax.inject.Singleton

/**
 * 鏡 Hub Manager — Discover and control Kagami Hub devices
 *
 * Colony: Nexus (e4) — Integration
 *
 * Features:
 * - mDNS/NSD service discovery
 * - Hub configuration and control
 * - Voice proxy (phone as Hub's ears)
 * - LED ring control
 * - Real-time status via WebSocket
 */

// ═══════════════════════════════════════════════════════════════════════════
// Data Classes
// ═══════════════════════════════════════════════════════════════════════════

@Serializable
data class HubDevice(
    val id: String,
    val name: String,
    val location: String,
    val host: String,
    val port: Int,
    val isConnected: Boolean = false
) {
    val baseUrl: String get() = "http://$host:$port"
}

@Serializable
data class HubStatus(
    val name: String,
    val location: String,
    val api_url: String,
    val api_connected: Boolean,
    val safety_score: Double? = null,
    val led_ring_enabled: Boolean,
    val led_brightness: Float,
    val wake_word: String,
    val is_listening: Boolean,
    val current_colony: String? = null,
    val uptime_seconds: Long,
    val version: String
)

@Serializable
data class HubConfig(
    val name: String,
    val location: String,
    val api_url: String,
    val wake_word: String,
    val wake_sensitivity: Float,
    val led_enabled: Boolean,
    val led_brightness: Float,
    val led_count: Int,
    val tts_volume: Float,
    val tts_colony: String
)

// ═══════════════════════════════════════════════════════════════════════════
// Hub Manager
// ═══════════════════════════════════════════════════════════════════════════

@Singleton
class HubManager @Inject constructor(
    @dagger.hilt.android.qualifiers.ApplicationContext private val context: Context
) {
    companion object {
        private const val TAG = "HubManager"
        private const val SERVICE_TYPE = "_kagami-hub._tcp."
        private const val PREFS_KEY = "kagami_hub_prefs"
        private const val LAST_HOST_KEY = "last_hub_host"
        private const val LAST_PORT_KEY = "last_hub_port"
    }

    private val json = Json { ignoreUnknownKeys = true }

    private val client = OkHttpClient.Builder()
        .connectTimeout(10, TimeUnit.SECONDS)
        .readTimeout(10, TimeUnit.SECONDS)
        .build()

    private val scope = CoroutineScope(Dispatchers.IO + SupervisorJob())

    // ═══════════════════════════════════════════════════════════════════════
    // State
    // ═══════════════════════════════════════════════════════════════════════

    private val _discoveredHubs = MutableStateFlow<List<HubDevice>>(emptyList())
    val discoveredHubs: StateFlow<List<HubDevice>> = _discoveredHubs.asStateFlow()

    private val _connectedHub = MutableStateFlow<HubDevice?>(null)
    val connectedHub: StateFlow<HubDevice?> = _connectedHub.asStateFlow()

    private val _hubStatus = MutableStateFlow<HubStatus?>(null)
    val hubStatus: StateFlow<HubStatus?> = _hubStatus.asStateFlow()

    private val _hubConfig = MutableStateFlow<HubConfig?>(null)
    val hubConfig: StateFlow<HubConfig?> = _hubConfig.asStateFlow()

    private val _isDiscovering = MutableStateFlow(false)
    val isDiscovering: StateFlow<Boolean> = _isDiscovering.asStateFlow()

    private val _isConnecting = MutableStateFlow(false)
    val isConnecting: StateFlow<Boolean> = _isConnecting.asStateFlow()

    private val _error = MutableStateFlow<String?>(null)
    val error: StateFlow<String?> = _error.asStateFlow()

    private var nsdManager: NsdManager? = null
    private var discoveryListener: NsdManager.DiscoveryListener? = null
    private var webSocket: WebSocket? = null

    // ═══════════════════════════════════════════════════════════════════════
    // Discovery via NSD (Android's mDNS)
    // ═══════════════════════════════════════════════════════════════════════

    fun startDiscovery() {
        if (_isDiscovering.value) return

        _isDiscovering.value = true
        _discoveredHubs.value = emptyList()
        _error.value = null

        nsdManager = context.getSystemService(Context.NSD_SERVICE) as NsdManager

        discoveryListener = object : NsdManager.DiscoveryListener {
            override fun onDiscoveryStarted(serviceType: String) {
                Log.d(TAG, "Discovery started for $serviceType")
            }

            override fun onServiceFound(serviceInfo: NsdServiceInfo) {
                Log.d(TAG, "Service found: ${serviceInfo.serviceName}")
                resolveService(serviceInfo)
            }

            override fun onServiceLost(serviceInfo: NsdServiceInfo) {
                Log.d(TAG, "Service lost: ${serviceInfo.serviceName}")
                _discoveredHubs.value = _discoveredHubs.value.filterNot {
                    it.name == serviceInfo.serviceName
                }
            }

            override fun onDiscoveryStopped(serviceType: String) {
                Log.d(TAG, "Discovery stopped")
                _isDiscovering.value = false
            }

            override fun onStartDiscoveryFailed(serviceType: String, errorCode: Int) {
                Log.e(TAG, "Discovery start failed: $errorCode")
                _error.value = "Discovery failed (error $errorCode)"
                _isDiscovering.value = false
            }

            override fun onStopDiscoveryFailed(serviceType: String, errorCode: Int) {
                Log.e(TAG, "Discovery stop failed: $errorCode")
            }
        }

        nsdManager?.discoverServices(SERVICE_TYPE, NsdManager.PROTOCOL_DNS_SD, discoveryListener)

        // Also try direct discovery
        scope.launch {
            tryDirectDiscovery()
        }

        // Stop discovery after 10 seconds
        scope.launch {
            delay(10_000)
            stopDiscovery()
        }
    }

    fun stopDiscovery() {
        discoveryListener?.let { listener ->
            try {
                nsdManager?.stopServiceDiscovery(listener)
            } catch (e: Exception) {
                Log.w(TAG, "Error stopping discovery: ${e.message}")
            }
        }
        discoveryListener = null
        _isDiscovering.value = false
    }

    private fun resolveService(serviceInfo: NsdServiceInfo) {
        nsdManager?.resolveService(serviceInfo, object : NsdManager.ResolveListener {
            override fun onServiceResolved(resolvedInfo: NsdServiceInfo) {
                val host = resolvedInfo.host?.hostAddress ?: return
                val port = resolvedInfo.port

                val hub = HubDevice(
                    id = "$host:$port",
                    name = resolvedInfo.serviceName,
                    location = "Discovered",
                    host = host,
                    port = port
                )

                scope.launch {
                    val current = _discoveredHubs.value.toMutableList()
                    if (current.none { it.id == hub.id }) {
                        current.add(hub)
                        _discoveredHubs.value = current
                    }
                }
            }

            override fun onResolveFailed(serviceInfo: NsdServiceInfo, errorCode: Int) {
                Log.e(TAG, "Resolve failed: $errorCode")
            }
        })
    }

    private suspend fun tryDirectDiscovery() {
        val candidates = listOf(
            "kagami-hub.local" to 8080,
            "raspberrypi.local" to 8080,
            "192.168.1.100" to 8080,
            "192.168.1.50" to 8080,
        )

        candidates.forEach { (host, port) ->
            if (testConnection(host, port)) {
                val hub = HubDevice(
                    id = "$host:$port",
                    name = "Kagami Hub",
                    location = "Direct",
                    host = host,
                    port = port
                )

                val current = _discoveredHubs.value.toMutableList()
                if (current.none { it.id == hub.id }) {
                    current.add(hub)
                    _discoveredHubs.value = current
                }
            }
        }
    }

    private suspend fun testConnection(host: String, port: Int): Boolean {
        return withContext(Dispatchers.IO) {
            try {
                val request = Request.Builder()
                    .url("http://$host:$port/health")
                    .build()

                client.newCall(request).execute().use { response ->
                    response.isSuccessful
                }
            } catch (e: Exception) {
                false
            }
        }
    }

    // ═══════════════════════════════════════════════════════════════════════
    // Connection
    // ═══════════════════════════════════════════════════════════════════════

    suspend fun connect(hub: HubDevice) {
        _isConnecting.value = true
        _error.value = null

        try {
            val status = fetchStatus(hub)
            val config = fetchConfig(hub)

            _connectedHub.value = hub.copy(
                isConnected = true,
                name = status.name,
                location = status.location
            )
            _hubStatus.value = status
            _hubConfig.value = config

            // Connect WebSocket
            connectWebSocket(hub)

            // Save as last hub
            saveLastHub(hub)

        } catch (e: Exception) {
            _error.value = "Connection failed: ${e.message}"
            Log.e(TAG, "Connection error", e)
        }

        _isConnecting.value = false
    }

    fun disconnect() {
        webSocket?.close(1000, "User disconnected")
        webSocket = null
        _connectedHub.value = null
        _hubStatus.value = null
        _hubConfig.value = null
    }

    // ═══════════════════════════════════════════════════════════════════════
    // API Calls
    // ═══════════════════════════════════════════════════════════════════════

    private suspend fun fetchStatus(hub: HubDevice): HubStatus {
        return withContext(Dispatchers.IO) {
            val request = Request.Builder()
                .url("${hub.baseUrl}/status")
                .build()

            client.newCall(request).execute().use { response ->
                if (!response.isSuccessful) throw IOException("Status fetch failed")
                json.decodeFromString(response.body!!.string())
            }
        }
    }

    private suspend fun fetchConfig(hub: HubDevice): HubConfig {
        return withContext(Dispatchers.IO) {
            val request = Request.Builder()
                .url("${hub.baseUrl}/config")
                .build()

            client.newCall(request).execute().use { response ->
                if (!response.isSuccessful) throw IOException("Config fetch failed")
                json.decodeFromString(response.body!!.string())
            }
        }
    }

    suspend fun updateConfig(config: HubConfig) {
        val hub = _connectedHub.value ?: return

        withContext(Dispatchers.IO) {
            val body = json.encodeToString(HubConfig.serializer(), config)
                .toRequestBody("application/json".toMediaType())

            val request = Request.Builder()
                .url("${hub.baseUrl}/config")
                .post(body)
                .build()

            client.newCall(request).execute().use { response ->
                if (!response.isSuccessful) throw IOException("Config update failed")
                _hubConfig.value = json.decodeFromString(response.body!!.string())
            }
        }
    }

    suspend fun controlLED(pattern: String, colony: Int? = null, brightness: Float? = null) {
        val hub = _connectedHub.value ?: return

        withContext(Dispatchers.IO) {
            val bodyJson = org.json.JSONObject().apply {
                put("pattern", pattern)
                colony?.let { put("colony", it) }
                brightness?.let { put("brightness", it.toDouble()) }
            }

            val body = bodyJson.toString().toRequestBody("application/json".toMediaType())

            val request = Request.Builder()
                .url("${hub.baseUrl}/led")
                .post(body)
                .build()

            client.newCall(request).execute().close()
        }
    }

    suspend fun testLED() {
        val hub = _connectedHub.value ?: return

        withContext(Dispatchers.IO) {
            val request = Request.Builder()
                .url("${hub.baseUrl}/led/test")
                .post("".toRequestBody())
                .build()

            client.newCall(request).execute().close()
        }
    }

    suspend fun triggerListen() {
        val hub = _connectedHub.value ?: return

        withContext(Dispatchers.IO) {
            val request = Request.Builder()
                .url("${hub.baseUrl}/voice/listen")
                .post("".toRequestBody())
                .build()

            client.newCall(request).execute().close()
        }
    }

    suspend fun executeCommand(command: String) {
        val hub = _connectedHub.value ?: return

        withContext(Dispatchers.IO) {
            val body = """{"command":"$command"}"""
                .toRequestBody("application/json".toMediaType())

            val request = Request.Builder()
                .url("${hub.baseUrl}/command")
                .post(body)
                .build()

            client.newCall(request).execute().close()
        }
    }

    // ═══════════════════════════════════════════════════════════════════════
    // WebSocket
    // ═══════════════════════════════════════════════════════════════════════

    private fun connectWebSocket(hub: HubDevice) {
        val request = Request.Builder()
            .url("ws://${hub.host}:${hub.port}/ws")
            .build()

        webSocket = client.newWebSocket(request, object : WebSocketListener() {
            override fun onOpen(webSocket: WebSocket, response: Response) {
                Log.d(TAG, "WebSocket connected")
            }

            override fun onMessage(webSocket: WebSocket, text: String) {
                handleWebSocketMessage(text)
            }

            override fun onFailure(webSocket: WebSocket, t: Throwable, response: Response?) {
                Log.e(TAG, "WebSocket error", t)
            }

            override fun onClosed(webSocket: WebSocket, code: Int, reason: String) {
                Log.d(TAG, "WebSocket closed: $reason")
            }
        })
    }

    private fun handleWebSocketMessage(text: String) {
        try {
            val event = json.decodeFromString<HubEvent>(text)

            when (event.event_type) {
                "status_update" -> {
                    // Parse nested status
                    _hubStatus.value?.let { current ->
                        // Update from event data
                    }
                }
                "listening_started" -> {
                    _hubStatus.value = _hubStatus.value?.copy(is_listening = true)
                }
                "config_updated" -> {
                    // Re-fetch config
                    scope.launch {
                        _connectedHub.value?.let { hub ->
                            _hubConfig.value = fetchConfig(hub)
                        }
                    }
                }
            }
        } catch (e: Exception) {
            Log.w(TAG, "Failed to parse WebSocket message", e)
        }
    }

    @Serializable
    private data class HubEvent(
        val event_type: String,
        val timestamp: Long
    )

    // ═══════════════════════════════════════════════════════════════════════
    // Persistence
    // ═══════════════════════════════════════════════════════════════════════

    private fun saveLastHub(hub: HubDevice) {
        context.getSharedPreferences(PREFS_KEY, Context.MODE_PRIVATE).edit()
            .putString(LAST_HOST_KEY, hub.host)
            .putInt(LAST_PORT_KEY, hub.port)
            .apply()
    }

    suspend fun connectToLastHub() {
        val prefs = context.getSharedPreferences(PREFS_KEY, Context.MODE_PRIVATE)
        val host = prefs.getString(LAST_HOST_KEY, null) ?: return
        val port = prefs.getInt(LAST_PORT_KEY, 8080)

        val hub = HubDevice(
            id = "$host:$port",
            name = "Last Hub",
            location = "Saved",
            host = host,
            port = port
        )

        connect(hub)
    }

    fun cleanup() {
        stopDiscovery()
        disconnect()
        scope.cancel()
    }
}

/*
 * 鏡
 */

File: orb/apps/ios/kagami-ios/KagamiIOS/Hub/HubManager.swift
//
// HubManager.swift — Hub Discovery and Control
//
// Colony: Nexus (e4) — Integration
//
// Discovers and manages Kagami Hub devices on the local network
// via Bonjour/mDNS and provides configuration and control.
//
// ## NWBrowser Memory Lifecycle
//
// The NWBrowser instance (`browser`) follows a strict lifecycle to prevent memory leaks:
//
// 1. **Creation**: Created in `startDiscovery()` with `NWBrowser(for:using:)`
// 2. **Start**: Started with `browser?.start(queue: .main)`
// 3. **Active**: Receives results via `browseResultsChangedHandler`
// 4. **Stop**: Must be cancelled with `browser?.cancel()` in `stopDiscovery()`
// 5. **Release**: Set to nil after cancellation to release memory
//
// ### Critical Rules:
// - ALWAYS call `browser?.cancel()` before setting browser to nil
// - ALWAYS call `stopDiscovery()` when:
//   - View disappears (e.g., in `.onDisappear`)
//   - Discovery timeout is reached
//   - User navigates away from Hub view
//   - App enters background
// - NEVER create a new browser without cancelling the old one first
//
// ### Memory Safety Pattern:
// ```swift
// func stopDiscovery() {
//     browser?.cancel()     // 1. Cancel first
//     browser = nil         // 2. Then release
//     isDiscovering = false
// }
// ```
//
// ### Connection Resolution:
// NWConnection instances created for service resolution are short-lived:
// - Created in `resolveService(result:name:)`
// - Cancelled immediately after endpoint extraction
// - Pattern: Create -> Start -> Extract -> Cancel
//
// h(x) >= 0. Always.
//

import Foundation
import Network
import Combine

// MARK: - Hub Models

struct HubDevice: Identifiable, Codable {
    let id: String
    var name: String
    var location: String
    var host: String
    var port: Int
    var isConnected: Bool
    var lastSeen: Date

    // Security: Hub communication uses HTTPS (self-signed certs on local network)
    var baseURL: String {
        "https://\(host):\(port)"
    }
}

struct HubStatus: Codable {
    let name: String
    let location: String
    let apiUrl: String
    let apiConnected: Bool
    let safetyScore: Double?
    let ledRingEnabled: Bool
    let ledBrightness: Float
    let wakeWord: String
    var isListening: Bool
    let currentColony: String?
    let uptimeSeconds: Int
    let version: String

    enum CodingKeys: String, CodingKey {
        case name, location, version
        case apiUrl = "api_url"
        case apiConnected = "api_connected"
        case safetyScore = "safety_score"
        case ledRingEnabled = "led_ring_enabled"
        case ledBrightness = "led_brightness"
        case wakeWord = "wake_word"
        case isListening = "is_listening"
        case currentColony = "current_colony"
        case uptimeSeconds = "uptime_seconds"
    }
}

struct HubConfig: Codable {
    var name: String
    var location: String
    var apiUrl: String
    var wakeWord: String
    var wakeSensitivity: Float
    var ledEnabled: Bool
    var ledBrightness: Float
    var ledCount: Int
    var ttsVolume: Float
    var ttsColony: String

    enum CodingKeys: String, CodingKey {
        case name, location
        case apiUrl = "api_url"
        case wakeWord = "wake_word"
        case wakeSensitivity = "wake_sensitivity"
        case ledEnabled = "led_enabled"
        case ledBrightness = "led_brightness"
        case ledCount = "led_count"
        case ttsVolume = "tts_volume"
        case ttsColony = "tts_colony"
    }
}

// MARK: - Hub Events (WebSocket)

struct HubEvent: Codable {
    let eventType: String
    let data: [String: AnyCodable]
    let timestamp: Int

    enum CodingKeys: String, CodingKey {
        case eventType = "event_type"
        case data, timestamp
    }
}

// Simple wrapper for heterogeneous JSON
public struct AnyCodable: Codable, @unchecked Sendable {
    public let value: Any

    public init(_ value: Any) {
        self.value = value
    }

    public init(from decoder: Decoder) throws {
        let container = try decoder.singleValueContainer()
        if let bool = try? container.decode(Bool.self) {
            value = bool
        } else if let int = try? container.decode(Int.self) {
            value = int
        } else if let double = try? container.decode(Double.self) {
            value = double
        } else if let string = try? container.decode(String.self) {
            value = string
        } else if let array = try? container.decode([AnyCodable].self) {
            value = array.map { $0.value }
        } else if let dict = try? container.decode([String: AnyCodable].self) {
            value = dict.mapValues { $0.value }
        } else {
            value = NSNull()
        }
    }

    public func encode(to encoder: Encoder) throws {
        var container = encoder.singleValueContainer()
        if let bool = value as? Bool {
            try container.encode(bool)
        } else if let int = value as? Int {
            try container.encode(int)
        } else if let double = value as? Double {
            try container.encode(double)
        } else if let string = value as? String {
            try container.encode(string)
        } else {
            try container.encodeNil()
        }
    }
}

// MARK: - Hub Manager

@MainActor
class HubManager: NSObject, ObservableObject {

    static let shared = HubManager()

    // MARK: - Published State

    @Published var discoveredHubs: [HubDevice] = []
    @Published var connectedHub: HubDevice?
    @Published var hubStatus: HubStatus?
    @Published var hubConfig: HubConfig?
    @Published var isDiscovering = false
    @Published var isConnecting = false
    @Published var lastError: String?

    // MARK: - Private

    private var browser: NWBrowser?
    private var webSocketTask: URLSessionWebSocketTask?
    private var session: URLSession

    // MARK: - Init

    override init() {
        let config = URLSessionConfiguration.default
        config.timeoutIntervalForRequest = 10
        self.session = URLSession(configuration: config)
        super.init()
    }

    // MARK: - Discovery via Bonjour

    func startDiscovery() {
        isDiscovering = true
        discoveredHubs.removeAll()

        // Create browser for _kagami-hub._tcp service
        let parameters = NWParameters()
        parameters.includePeerToPeer = true

        browser = NWBrowser(for: .bonjour(type: "_kagami-hub._tcp", domain: "local."), using: parameters)

        browser?.stateUpdateHandler = { [weak self] state in
            Task { @MainActor in
                switch state {
                case .ready:
                    print("🔍 Hub discovery ready")
                case .failed(let error):
                    self?.lastError = "Discovery failed: \(error)"
                    self?.isDiscovering = false
                default:
                    break
                }
            }
        }

        browser?.browseResultsChangedHandler = { [weak self] results, changes in
            Task { @MainActor in
                self?.handleDiscoveryResults(results)
            }
        }

        browser?.start(queue: .main)

        // Also try direct connection to common addresses
        Task {
            await tryDirectDiscovery()
        }

        // Stop discovery after 10 seconds
        Task {
            try? await Task.sleep(nanoseconds: 10_000_000_000)
            await MainActor.run {
                self.stopDiscovery()
            }
        }
    }

    func stopDiscovery() {
        browser?.cancel()
        browser = nil
        isDiscovering = false
    }

    private func handleDiscoveryResults(_ results: Set<NWBrowser.Result>) {
        for result in results {
            if case .service(let name, _, _, _) = result.endpoint {
                // Resolve the service to get host/port
                resolveService(result: result, name: name)
            }
        }
    }

    private func resolveService(result: NWBrowser.Result, name: String) {
        let connection = NWConnection(to: result.endpoint, using: .tcp)

        connection.stateUpdateHandler = { [weak self] state in
            if case .ready = state {
                if let endpoint = connection.currentPath?.remoteEndpoint,
                   case .hostPort(let host, let port) = endpoint {
                    let hostString = "\(host)"
                    let portInt = Int(port.rawValue)

                    Task { @MainActor [weak self] in
                        guard let self = self else { return }
                        let hub = HubDevice(
                            id: "\(hostString):\(portInt)",
                            name: name,
                            location: "Discovered",
                            host: hostString,
                            port: portInt,
                            isConnected: false,
                            lastSeen: Date()
                        )

                        if !self.discoveredHubs.contains(where: { $0.id == hub.id }) {
                            self.discoveredHubs.append(hub)
                        }
                    }
                }
                connection.cancel()
            }
        }

        connection.start(queue: .global())
    }

    private func tryDirectDiscovery() async {
        // Try common Hub addresses
        let candidates = [
            ("kagami-hub.local", 8080),
            ("raspberrypi.local", 8080),
            ("192.168.1.100", 8080),
            ("192.168.1.50", 8080),
        ]

        for (host, port) in candidates {
            if await testHubConnection(host: host, port: port) {
                let hub = HubDevice(
                    id: "\(host):\(port)",
                    name: "Kagami Hub",
                    location: "Direct",
                    host: host,
                    port: port,
                    isConnected: false,
                    lastSeen: Date()
                )

                await MainActor.run {
                    if !discoveredHubs.contains(where: { $0.id == hub.id }) {
                        discoveredHubs.append(hub)
                    }
                }
            }
        }
    }

    private func testHubConnection(host: String, port: Int) async -> Bool {
        // Security: Use HTTPS for hub health checks
        guard let url = URL(string: "https://\(host):\(port)/health") else { return false }

        do {
            let (_, response) = try await session.data(from: url)
            return (response as? HTTPURLResponse)?.statusCode == 200
        } catch {
            return false
        }
    }

    // MARK: - Connection

    func connect(to hub: HubDevice) async {
        isConnecting = true
        lastError = nil

        do {
            // Fetch status
            let status = try await fetchStatus(from: hub)
            let config = try await fetchConfig(from: hub)

            var connectedHub = hub
            connectedHub.isConnected = true
            connectedHub.name = status.name
            connectedHub.location = status.location

            self.connectedHub = connectedHub
            self.hubStatus = status
            self.hubConfig = config

            // Connect WebSocket for real-time updates
            connectWebSocket(to: hub)

            // Save as last connected hub
            saveLastHub(hub)

        } catch {
            lastError = "Connection failed: \(error.localizedDescription)"
        }

        isConnecting = false
    }

    func disconnect() {
        webSocketTask?.cancel(with: .goingAway, reason: nil)
        webSocketTask = nil
        connectedHub = nil
        hubStatus = nil
        hubConfig = nil
    }

    // MARK: - API Calls

    private func fetchStatus(from hub: HubDevice) async throws -> HubStatus {
        guard let url = URL(string: "\(hub.baseURL)/status") else {
            throw URLError(.badURL)
        }

        let (data, _) = try await session.data(from: url)
        return try JSONDecoder().decode(HubStatus.self, from: data)
    }

    private func fetchConfig(from hub: HubDevice) async throws -> HubConfig {
        guard let url = URL(string: "\(hub.baseURL)/config") else {
            throw URLError(.badURL)
        }

        let (data, _) = try await session.data(from: url)
        return try JSONDecoder().decode(HubConfig.self, from: data)
    }

    func updateConfig(_ config: HubConfig) async throws {
        guard let hub = connectedHub,
              let url = URL(string: "\(hub.baseURL)/config") else {
            throw URLError(.badURL)
        }

        var request = URLRequest(url: url)
        request.httpMethod = "POST"
        request.setValue("application/json", forHTTPHeaderField: "Content-Type")
        request.httpBody = try JSONEncoder().encode(config)

        let (data, _) = try await session.data(for: request)
        self.hubConfig = try JSONDecoder().decode(HubConfig.self, from: data)
    }

    func controlLED(pattern: String, colony: Int? = nil, brightness: Float? = nil) async throws {
        guard let hub = connectedHub,
              let url = URL(string: "\(hub.baseURL)/led") else {
            throw URLError(.badURL)
        }

        var body: [String: Any] = ["pattern": pattern]
        if let colony = colony { body["colony"] = colony }
        if let brightness = brightness { body["brightness"] = brightness }

        var request = URLRequest(url: url)
        request.httpMethod = "POST"
        request.setValue("application/json", forHTTPHeaderField: "Content-Type")
        request.httpBody = try JSONSerialization.data(withJSONObject: body)

        let _ = try await session.data(for: request)
    }

    func testLED() async throws {
        guard let hub = connectedHub,
              let url = URL(string: "\(hub.baseURL)/led/test") else {
            throw URLError(.badURL)
        }

        var request = URLRequest(url: url)
        request.httpMethod = "POST"

        let _ = try await session.data(for: request)
    }

    func triggerListen() async throws {
        guard let hub = connectedHub,
              let url = URL(string: "\(hub.baseURL)/voice/listen") else {
            throw URLError(.badURL)
        }

        var request = URLRequest(url: url)
        request.httpMethod = "POST"

        let _ = try await session.data(for: request)
    }

    func executeCommand(_ command: String) async throws {
        guard let hub = connectedHub,
              let url = URL(string: "\(hub.baseURL)/command") else {
            throw URLError(.badURL)
        }

        var request = URLRequest(url: url)
        request.httpMethod = "POST"
        request.setValue("application/json", forHTTPHeaderField: "Content-Type")
        request.httpBody = try JSONSerialization.data(withJSONObject: ["command": command])

        let _ = try await session.data(for: request)
    }

    // MARK: - WebSocket

    private func connectWebSocket(to hub: HubDevice) {
        guard let url = URL(string: "ws://\(hub.host):\(hub.port)/ws") else { return }

        webSocketTask?.cancel()
        webSocketTask = session.webSocketTask(with: url)
        webSocketTask?.resume()

        receiveWebSocketMessage()
    }

    private func receiveWebSocketMessage() {
        webSocketTask?.receive { [weak self] result in
            switch result {
            case .success(let message):
                if case .string(let text) = message {
                    self?.handleWebSocketMessage(text)
                }
                self?.receiveWebSocketMessage()

            case .failure(let error):
                print("WebSocket error: \(error)")
            }
        }
    }

    private func handleWebSocketMessage(_ text: String) {
        guard let data = text.data(using: .utf8),
              let event = try? JSONDecoder().decode(HubEvent.self, from: data) else { return }

        Task { @MainActor in
            switch event.eventType {
            case "status_update":
                if let statusData = try? JSONSerialization.data(withJSONObject: event.data.mapValues { $0.value }),
                   let status = try? JSONDecoder().decode(HubStatus.self, from: statusData) {
                    self.hubStatus = status
                }

            case "listening_started":
                self.hubStatus?.isListening = true

            case "config_updated":
                if let configData = try? JSONSerialization.data(withJSONObject: event.data.mapValues { $0.value }),
                   let config = try? JSONDecoder().decode(HubConfig.self, from: configData) {
                    self.hubConfig = config
                }

            default:
                break
            }
        }
    }

    // MARK: - Persistence

    private func saveLastHub(_ hub: HubDevice) {
        UserDefaults.standard.set(hub.host, forKey: "lastHubHost")
        UserDefaults.standard.set(hub.port, forKey: "lastHubPort")
    }

    func connectToLastHub() async {
        guard let host = UserDefaults.standard.string(forKey: "lastHubHost") else { return }
        let port = UserDefaults.standard.integer(forKey: "lastHubPort")

        let hub = HubDevice(
            id: "\(host):\(port)",
            name: "Last Hub",
            location: "Saved",
            host: host,
            port: port > 0 ? port : 8080,
            isConnected: false,
            lastSeen: Date()
        )

        await connect(to: hub)
    }
}

/*
 * 鏡
 */
